wai-logger/Network/Wai/Logger/Apache.hs

```diff
@@ -154,15 +189,65 @@ getSourceFromFallback req = fromMaybe (getSourceFromSocket req) $ getSource req
+-- Just "5.6.7.8"
+--
+-- 'getSource' handles pulling out the first IP in the comma-separated IP list
+-- in X-Forwarded-For:
+--
+-- >>> getSource defaultRequest { requestHeaders = [ ("X-Forwarded-For", "5.6.7.8, 10.11.12.13, 1.2.3.4") ] }
+-- Just "5.6.7.8"
 getSource :: Request -> Maybe ByteString
-getSource = getSourceFromHeaders ["x-real-ip", "x-forwarded-for"]
+getSource = getSourceFromHeaders [("x-real-ip", id), ("x-forwarded-for", firstIpInXFF)]
 
-getSourceFromHeaders :: [HeaderName] -> Request -> Maybe ByteString
-getSourceFromHeaders headerNames req = addr
+-- | Pull out the first IP in a comma-separated list of X-Forwarded-For IPs.
+--
+-- >>> firstIpInXFF "1.2.3.4, 5.6.7.8, 10.11.12.13"
+-- "1.2.3.4"
+--
+-- If there are no commas, just return the whole input ByteString:
+--
+-- >>> firstIpInXFF "5.6.7.8"
+-- "5.6.7.8"
+--
+-- Note that this function doesn't make sure the input is actually an IP address:
+--
+-- >>> firstIpInXFF "hello, world"
+-- "hello"
+firstIpInXFF :: ByteString -> ByteString
+firstIpInXFF = BS.takeWhile (/= ',')
+
+getSourceFromHeaders :: [(HeaderName, ByteString -> ByteString)] -> Request -> Maybe ByteString
+getSourceFromHeaders headerNamesAndPostProc req = getFirst $ foldMap f $ requestHeaders req
   where
-    maddr = find (\(name,_) -> name `elem` headerNames) hdrs
-    addr = fmap snd maddr
-    hdrs = requestHeaders req
+    -- Take a header name and value from the request, and try match it against
+    -- the list of headers and post-processing functions.  If it matches,
+    -- return the ByteString resulting from applying the post-processing function
+    -- to the header value.
+    f :: (HeaderName, ByteString) -> First ByteString
+    f (headerNameFromReq, headerValFromReq) =
+      let maybePostProc = find (\(headerNameFromPostProc, _) -> headerNameFromReq == headerNameFromPostProc) headerNamesAndPostProc
+      in First $ fmap (\(_, postProc) -> postProc headerValFromReq) maybePostProc
 
+-- |
+-- >>> getSourceFromHeaderCustom ["x-foobar"] defaultRequest { requestHeaders = [ ("X-catdog", "1.2.3.4"), ("X-Foobar", "5.6.7.8"), ("Other", "1.1.1.1") ] }
+-- Just "5.6.7.8"
+--
+-- If none of the headers in the passed-in list are in the 'Request', then return 'Nothing':
+--
+-- >>> getSourceFromHeaderCustom ["x-foobar", "baz"] defaultRequest { requestHeaders = [ ("abb", "1.2.3.4"), ("xyz", "5.6.7.8") ] }
+-- Nothing
+--
+-- 'getSourceFromHeaderCustom' uses the first instance of any header in the
+-- passed in list that it finds in the ordered header list from the request:
+--
+-- >>> getSourceFromHeaderCustom ["x-foobar", "baz"] defaultRequest { requestHeaders = [ ("baz", "1.2.3.4"), ("x-foobar", "5.6.7.8") ] }
+-- Just "1.2.3.4"
 getSourceFromHeaderCustom :: [HeaderName] -> Request -> Maybe ByteString
-getSourceFromHeaderCustom hs = getSourceFromHeaders hs
+getSourceFromHeaderCustom hs = getSourceFromHeaders (fmap (,id) hs)
```

Comment:
Can we use the `firstIpInXFF` here as well? The case would be when you need "x-forwarded-for" but you cannot use getSource because it would read "x-real-ip" first when both are present.

Reply:
We could use `firstIpInXFF` in `getSourceFromHeaderCustom`. However, my thought is that it would be somewhat surprising for a user, so I didn't end up adding it.

Really what would be nice is a new data constructor like:

```haskell
data IPAddrSource
  = ...
  | FromRequest (Request -> ByteString)
```

so that the end-user could be in full control of exactly how to figure out the IP address.  I didn't choose to add that in this PR, but I could if there is a desire to have it.

Reply:
What about both using `firstIpInXFF` in `getSourceFromHeaderCustom` *AND* adding `FromRequest` above?

Reply:
Not a bad idea, this FromRequest supersedes all the other options, but it may be nice to have a choice.

Reply:
@cdepillabout I you agree, please push a commit to implement the two things above.

Reply:
@kazu-yamamoto I pushed a couple commits, with the important ones being:

- e349e591b5dd: Add the `FromRequest (Request -> ByteString)` data constructor.
- 1d519c6bf9f1: Use `firstIpInXFF` in `getSourceFromHeaderCustom`.  I made the choice to apply `firstIpInXFF` to all the headers passed in to `getSourceFromHeaderCustom`, since it should effectively be a no-op if the header value doesn't have a comma in it.
- 2d2563240dcd7: Bump wai-logger to 2.5.0.  (Feel free to drop this commit if you'd prefer to do this yourself in a separate PR or something)